### DISCUSSION TITLE
Is the Webstudio builder open source or an I reading it wrong?

### DISCUSSION THREAD

**Discussion by codewulf0** [webstudio-is/webstudio-community#150] [Q&A] [answered]
Hello, I'm confused about the use of the term open source in Webstudio's communication. On the website it mentions that the Open Source Builder is free, but when I look on Github I only see explanations for how to build a Project that was designed on Cloud. What I was looking for is the Cloud designer on a self-hosted solution, not the final site that can be build locally. Am I misinterpreting things, or am I missing a key piece of information? Sorry if this has been asked before, I looked around but didnt find the anwser.

**Comment by kof** [comment, marked as answer]
It is in fact open source, AGPL licensed.  What you are missing is the grasp of the architecture for hosting. Webstudio builder is built on top of a number of services.

TLDR: We haven't made it easy to self-host using docker image yet and publish & build on the same machine. We have just recently done the first steps towards this (devcontainers) and will be able to enable a self-hostable container soon. That being said we have higher priority issues and a small team, so if anyone wants to do this - it would speed it up. We are happy to answer any questions.


List of things to consider:

- Build pipeline  - what happens when you hit publish (we currently use github actions to build the app using the CLI)
- Publishing pipeline - depending on how you want to host production sites, you need different strategies, we deploy to cloudflare worker for platforms, its something only we as a cloud provider need. As a consumer you would need to either decide on a different cloud (e.g. vercel, netlify, cloudflare pages) or use a self-hostable container, in which case you would be hosting both builder and the site on the same machine or you would go for some other custom strategy for publishing.
- Image resizer and converter - Cloudflare Image resizer
- AI features depend on OpenAI and Pexels
- Auth depends on Google Auth and Github Auth
- Database - Postgres, we are using supabase

**Reply by codewulf0** [reply]
Thanks for your reply Kof, you're right that the concept of the hosting architecture is still a far cry for me. I've taught myself how to run Docker containers, and have managed to get several services running in Railway.app (and would love to run my own Webstudio builder there). But I fear I still lack several key concepts.

If by chance someone would have 30 minutes to set up a Railway.app enviornment (if at all possible), we could make that a public template on Railway and allow more people to run it. But only if it's possible given the current dev. stage ofcourse.

**Reply by BenAlanSouthall** [reply]
I guess even if not perfect, it'd definitely help to have a bit more detail about what is there currently and be nice to try it out. I don't know about the database as it depends on how the backend is written - but for my scenario, the other bullet points aren't that important. I can convert images locally, I don't care about authentication as it'd only be served on localhost, and the deployment I can do by hand if I can get a hold of a static export. So point being whilst I know some projects don;t like to release things too early as it can be harder to maintain projects once they're out in the wild than within an internal team, on the flip side some basic WIP instructions for a not-ready-yet setup are probably an improvement on waiting until all the loose ends are tied up. And hopefully then that gives you more users who can feed back issues with self-hosting. So IMO would be nice to be able to try it out with limited functionality and can feed back on any issues :)

**Reply by kof** [reply]
Just do it https://docs.webstudio.is/university/self-hosting

**Reply by BenAlanSouthall** [reply]
That's what I'm asking though - is there WIP simple container image/Dockerfile I can pull down or build to run it? Or is that not yet pushed up to the repository? The main docs section contains the warning that self-hosting is not recommended but then no instructions, and then the link goes to the development workflow, making code changes, and using a remote workspace in github etc, which is a bit confusing to me. I'd just like to be able to run it and access it on localhost if possible. If that's not ready yet, I can at least try it and see if I can do anything to improve the container recipe

**Reply by kof** [reply]
We have a container we use for development and its linked in the docs. Its not recommended for production as its not complete for production use cases like publishing etc. Its still a process that requires running CLI to build and publish. In our cloud version we use github actions to build and publish using the same CLI, but in docker there is no action runners and few other things https://github.com/webstudio-is/webstudio/issues/3966

**Reply by kof** [reply]
To try things out you can certainly use our docker container that is referenced in the docs for developer contribution.

**Reply by stellarpower** [reply]
I don't see any images or tags mentioned [here]( https://docs.webstudio.is/contributing/contributing-for-developers ), also searched the docs for "docker". Is it `rogerznts/docker-webstudio:dev` (from [here](  https://github.com/rogerznts/webstudio/blob/feature/docker-compose/docker/docker-compose.yml#L8  ))? If so, it'd help to have it linked on the self-hosting page under the "Builder" section. Execcing in and using the CLI is perfectly fine for me - probably more convenient actually, as I can automate it. It'd be easier to export the site through a volume and publish myself than need to set up all the authentication inside the container. So not a problem :)

EDIT: accidentally replied from my other account.

**Reply by TrySound** [reply]
https://github.com/webstudio-is/webstudio/blob/main/.devcontainer/docker-compose.yml

**Reply by johnsicili** [reply]
> I don't see any images or tags mentioned [here](https://docs.webstudio.is/contributing/contributing-for-developers),

We have not released any docker images, the only good way to execute is to use devcontainer (which is docker-compose but with additional steps https://containers.dev/ )

**Reply by stellarpower** [reply]
Okay, thanks a lot for confirming. I'll try building it from that Dockerfile and see what happens.

**Reply by stellarpower** [reply]
Can I ask the reason it currently needs docker-in-docker?

**Reply by istarkov** [reply]
https://github.com/webstudio-is/webstudio/blame/main/.devcontainer/Dockerfile

**Reply by stellarpower** [reply]
I see Supabase CLI mentioned, but googling it, it can be installed using npm or distribution packages. I assume there must be a motivating reason, but I expect if it can be avoided, it would probably be easier to improve and maintain using a straightforward pod approach.

Also I get a 502 when trying to open the redirected port using github's codespace. Is this a known situation?

**Reply by istarkov** [reply]
If you run the Supabase CLI installed through npm, you'll notice that it executes a Docker container and won't work without Docker installed. This shows that some things can't be discovered just by Googling; they exist in the real world.

**Comment by vianmora** [comment]
I've created a repo with a doc for self-hosting the builder if you want :) Letting the webstudio team focus improve the builder with wonderfull features :) 

-> https://github.com/webstudio-community/webstudio-self-host
